File: 3740-minimum-distance-between-three-equal-elements-i/3740-minimum-distance-between-three-equal-elements-i.go
func minimumDistance(nums []int) int {
    n := len(nums)
    if n < 3 {
        return -1
    }

	x := int(10e10)
    exists := false
	for i := 0; i < n-2; i++ {
		for j := i + 1; j < n-1; j++ {
			for k := j + 1; k < n; k++ {
				if nums[i] == nums[j] && nums[j] == nums[k] && nums[i] == nums[k] {
					x = min(x, abs(i-j)+abs(j-k)+abs(k-i))
                    exists = true
				}
			}
		}
	}

	if exists {
        return x
    }
    return -1
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

// array
// time: O(n^3)
// space: O(1)

File: 3740-minimum-distance-between-three-equal-elements-i/3740-minimum-distance-between-three-equal-elements-i.py
class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        n = len(nums)
        if n < 3:
            return -1

        m = {}
        x = 10**10
        found = False
        for i in range(n):
            if nums[i] not in m:
                m[nums[i]] = [i]
            else:
                m[nums[i]].append(i)
                nn = len(m[nums[i]])
                if nn >= 3:
                    a, b, c = m[nums[i]][nn - 3], m[nums[i]][nn - 2], m[nums[i]][nn - 1]
                    x = min(x, abs(a - b) + abs(b - c) + abs(c - a))
                    found = True
        
        if found:
            return x
        return -1

# array, hash map
# time: O(n)
# space: O(n)